js/ui/appDisplay.js

```diff
@@ -167,23 +174,193 @@ const BaseAppView = new Lang.Class({
+        let newItemLayout = this.getLayoutIds();
+        newItemLayout = this._trimInvisible(newItemLayout);
+
+        let movedList = {};
+        let removedList = [];
+        for (let oldItemIdx in oldItemLayout) {
+            let oldItem = oldItemLayout[oldItemIdx];
+            let newItemIdx = newItemLayout.indexOf(oldItem);
+
+            if (oldItemIdx != newItemIdx) {
+                if (newItemIdx < 0) {
+                    removedList.push(oldItemIdx);
+                } else {
+                    movedList[oldItemIdx] = newItemIdx;
+                }
+            }
+        }
+
+        return [movedList, removedList];
+    },
+
+    _findAddedIcons: function() {
+        let oldItemLayout = this._allItems.map(function(icon) { return icon.getId(); });
+        if (oldItemLayout.length === 0) return [];
+
+        let newItemLayout = this.getLayoutIds();
+        newItemLayout = this._trimInvisible(newItemLayout);
+
+        let addedIds = [];
+        for (let newItemIdx in newItemLayout) {
+            let newItem = newItemLayout[newItemIdx];
+            let oldItemIdx = oldItemLayout.indexOf(newItem);
+
+            if (oldItemIdx < 0) {
+                addedIds.push(newItem);
+            }
+        }
+
+        return addedIds;
+    },
+
+    animateMovement: function() {
+        let [movedList, removedList] = this._findIconChanges();
+        this._grid.animateShuffling(movedList,
+                                    removedList,
+                                    this.repositionedIconData,
+                                    Lang.bind(this, this.addIcons));
+        this.repositionedIconData = [ null, null ];
+    },
+
+    iconsNeedRedraw: function() {
+        // Check if the icons moved around
+        let [movedList, removedList] = this._findIconChanges();
+        let movedLength = Object.keys(movedList).length;
+        if (movedLength > 0 || removedList.length > 0)
+            return true;
+
+        // Create a map from app ids to icon objects
+        let iconTable = {};
+        for (let idx in this._allItems)
+            iconTable[this._allItems[idx].getId()] = this._allItems[idx];
+
+        let layoutIds = this.getLayoutIds();
+
+        // Iterate through all visible icons
+        for (let idx in layoutIds) {
+            let itemId = layoutIds[idx];
+
+            let item = this._createItemForId(itemId);
+            if (!item)
+                continue;
+
+            // The App Center icon cannot be changed or renamed
+            if (item == this._appCenterItem)
+                continue;
+
+            let currentIcon = iconTable[itemId];
+            if (!currentIcon)
+                return true; // This icon is new
+
+            let isFolder = IconGridLayout.layout.iconIsFolder(itemId);
+            let oldIconInfo = null;
+            let newIconInfo = null;
+
+            if (isFolder) {
+                oldIconInfo = currentIcon.folder.get_icon();
```

Comment:
Do we have a generic `getIcon()` method that we could call on the parent item class here?

Reply:
Hm, looks like we don't. There are `getId` and `getName`, but no `getIcon`.

Reply:
Could you add that method?

Reply:
+1 to Cosimo's comment

Reply:
Implemented this change here: https://github.com/endlessm/gnome-shell/compare/ba1bc99fced9195a3d086a9ce153a64cbab83edf...331b457118abb801384f0c696c8b21044a235c45